### DISCUSSION TITLE
NVIDIA Jetson and Create 3

### DISCUSSION THREAD

**Discussion by eecue** [iRobotEducation/create3_docs#73] [Questions] [answered]
I followed the doc here and everything works as described except I am unable to get the topics, instead I just see:

```
/parameter_events
/rosout
```

My networking looks right:

```
l4tbr0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.186.3  netmask 255.255.255.0  broadcast 192.168.186.255
        inet6 fe80::1  prefixlen 128  scopeid 0x20<link>
        inet6 fe80::bc07:b3ff:fe27:44f1  prefixlen 64  scopeid 0x20<link>
        ether be:07:b3:27:44:f1  txqueuelen 1000  (Ethernet)
        RX packets 164  bytes 53280 (53.2 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 353  bytes 96957 (96.9 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

What's strange is that if I nmap that IP I just see the services my Jetson is running.

**Comment by eecue** [comment]
I'm wondering if it's my USB cable... trying a different one.

**Comment by eecue** [comment]
Hmm don't think that was it, when I pulled the cable I saw this in the logs:

``` 
Jun  6 21:53:13 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [14750.540057] usb 1-1.2.3: USB disconnect, device number 8
Jun  6 21:53:13 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [14750.581823] cdc_ncm 1-1.2.3:1.5 usb0: unregister 'cdc_ncm' usb-101c0000.ehci-1.2.3, CDC NCM
Jun  6 21:53:14 iRobot-06C0C83B61B040079D77F8D529A014C2 daemon.info avahi-daemon[1239]: Interface usb0.IPv4 no longer relevant for mDNS.
Jun  6 21:53:14 iRobot-06C0C83B61B040079D77F8D529A014C2 daemon.info avahi-daemon[1239]: Leaving mDNS multicast group on interface usb0.IPv4 with address 192.168.186.2.
Jun  6 21:53:14 iRobot-06C0C83B61B040079D77F8D529A014C2 daemon.info avahi-daemon[1239]: Withdrawing address record for 192.168.186.2 on usb0.
```

**Reply by valdivj** [reply]
I tried setting up my Xavier NX to use the cable. But once i put the create3 on my wifi network all the ROS2 commands worked. So I gave up on the cable. I really didn't do anything extra too get it to accept commands once it was on my wifi network

**Comment by eecue** [comment]
Here's what I see when I plug it back in:

```
Jun  6 21:57:53 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [15029.867767] usb 1-1.2.4: new high-speed USB device number 9 using ehci-platform
Jun  6 21:57:53 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [15030.019545] usb 1-1.2.4: New USB device found, idVendor=0955, idProduct=7020
Jun  6 21:57:53 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [15030.019567] usb 1-1.2.4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
Jun  6 21:57:53 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [15030.019577] usb 1-1.2.4: Product: Linux for Tegra
Jun  6 21:57:53 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [15030.019586] usb 1-1.2.4: Manufacturer: NVIDIA
Jun  6 21:57:53 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [15030.019596] usb 1-1.2.4: SerialNumber: 1423019047978
Jun  6 21:57:53 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [15030.030543] usb-storage 1-1.2.4:1.4: USB Mass Storage device detected
Jun  6 21:57:53 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [15030.051444] scsi host0: usb-storage 1-1.2.4:1.4
Jun  6 21:57:53 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [15030.089589] cdc_ncm 1-1.2.4:1.5: MAC-Address: be:07:b3:27:44:f2
Jun  6 21:57:53 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [15030.090612] cdc_ncm 1-1.2.4:1.5 usb0: register 'cdc_ncm' at usb-101c0000.ehci-1.2.4, CDC NCM, be:07:b3:27:44:f2
Jun  6 21:57:53 iRobot-06C0C83B61B040079D77F8D529A014C2 daemon.info avahi-daemon[1239]: Joining mDNS multicast group on interface usb0.IPv4 with address 192.168.186.2.
Jun  6 21:57:53 iRobot-06C0C83B61B040079D77F8D529A014C2 daemon.info avahi-daemon[1239]: New relevant interface usb0.IPv4 for mDNS.
Jun  6 21:57:53 iRobot-06C0C83B61B040079D77F8D529A014C2 daemon.info avahi-daemon[1239]: Registering new address record for 192.168.186.2 on usb0.IPv4.
Jun  6 21:57:53 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [15030.700951] cdc_ncm 1-1.2.4:1.5 usb0: 425 mbit/s downlink 425 mbit/s uplink
Jun  6 21:57:54 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [15030.732943] cdc_ncm 1-1.2.4:1.5 usb0: 425 mbit/s downlink 425 mbit/s uplink
Jun  6 21:57:54 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [15030.764941] cdc_ncm 1-1.2.4:1.5 usb0: network connection: connected
Jun  6 21:57:54 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.err kernel: [15030.765996] cdc_ncm 1-1.2.4:1.5 usb0: kevent 2 may have been dropped
Jun  6 21:57:54 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.notice kernel: [15031.083857] scsi 0:0:0:0: Direct-Access     Linux    File-Stor Gadget 0409 PQ: 0 ANSI: 2
Jun  6 21:57:54 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.notice kernel: [15031.089252] sd 0:0:0:0: [sda] 32768 512-byte logical blocks: (16.8 MB/16.0 MiB)
Jun  6 21:57:54 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.notice kernel: [15031.090165] sd 0:0:0:0: [sda] Write Protect is on
Jun  6 21:57:54 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.debug kernel: [15031.090220] sd 0:0:0:0: [sda] Mode Sense: 0f 00 80 00
Jun  6 21:57:54 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.notice kernel: [15031.091541] sd 0:0:0:0: [sda] Write cache: disabled, read cache: enabled, doesn't support DPO or FUA
Jun  6 21:57:54 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.info kernel: [15031.104265]  sda:
Jun  6 21:57:54 iRobot-06C0C83B61B040079D77F8D529A014C2 kern.notice kernel: [15031.123405] sd 0:0:0:0: [sda] Attached SCSI removable disk
```

**Comment by shamlian** [comment]
Looks to me like Create is seeing the Jetson as it should, and IP addresses look correct. Can you ping the Create from the Jetson? If not, it might be worth trying nmtui to reconfigure the connection on the Jetson, as with [this post](https://github.com/iRobotEducation/create3_docs/discussions/59#discussioncomment-2780077).

Also -- maybe a long shot, but worth checking if you can ping the Create from the Jetson: are both the Jetson and the Create using the same middleware (either Cyclone or Fast-DDS)? I have made this mistake more times than I care to admit...

I have a Jetson Nano 2GB and will attempt a run-through of the tutorial tomorrow to see if I can reproduce your problem.

**Reply by eecue** [reply]
I can’t ping it. I’ll try this as soon as I get home!

**Reply by eecue** [reply]
I tried this and it didn’t work.

**Comment by eecue** [comment]
I still can't ping it, crazy thing though is that if I run tcpdump on that interface I see traffic from the Create:

`tcpdump -i l4tbr0 -X -s 5000 -vvv`

```
18:13:52.552587 IP (tos 0x0, ttl 1, id 3180, offset 0, flags [DF], proto UDP (17), length 336)
    192.168.186.2.32914 > 239.255.0.1.7400: [udp sum ok] UDP, length 308
```

**Comment by wpd** [comment]
Is it possible that you are using IPv6's ping (ping6) instead of IPv4 ping?  I'm not familiar with the Jetson environment, but I'm great at inventing theories 🙂
Expanding a little bit... 
Perhaps `ping` is aliased to `ping6` on the Jetson and you could try something like:
`ping -4 192.168.186.2`
although, if that makes things work, that would make me sad... since I would have hoped that `ping` would recognize that you wanted to use IPv4 by the fact that you specified a dotted quad in the first place :-(

**Reply by eecue** [reply]
Nope that didn’t do it. I can ping other IPv4 hosts on my network. And yes there for sure is a route to the 192.168.186.0 via l4tbr0

**Reply by shamlian** [reply]
@tokk-nv  -- any ideas?

**Comment by shamlian** [comment]
I'm not sure what your setup looks like, but this morning I set up a Jetson Nano 2GB dev kit to successfully ping a Create 3 over its USB C (cdcncm) port by doing the following:

1. Perform headless initial setup, as discussed on [NVIDIA's site](https://developer.nvidia.com/embedded/learn/get-started-jetson-nano-2gb-devkit#setup). One important point is that after the `dummy0` network interface setup failed, I told the setup wizard that I was not interested in setting up the network connection at that point.
2. Perform step 1 of the directions on the [docs site](https://iroboteducation.github.io/create3_docs/setup/jetson/) -- first, copy and paste step 1a, and then completely replace `/opt/nvidia/l4t-usb-device-mode/nv-l4t-usb-device-mode-runtime-start.sh` with the file provided in step 1b.
3. `sudo shutdown now` on the Jetson. At this point, I connected the Jetson to my local wired Ethernet in order to verify things were working. I don't think this is strictly necessary, but otherwise there is no other available interface for me to connect to my Jetson headless.
4. Plug the USB Micro-B and USB C connectors into the Nano via a hub into the Create 3 as shown [here](https://iroboteducation.github.io/create3_docs/hw/hookup/#nvidia-jetson-nanotm-2gb).
5. Wait for the Jetson to boot, and then connect to it (in my case, over wired Ethernet). At this point, performing step 2 of the directions on the [docs site](https://iroboteducation.github.io/create3_docs/setup/jetson/) (`ifconfig l4tbr0`) shows the connection as expected, and ping also returns packets.

**Reply by shamlian** [reply]
Just for apples to apples, my Create 3 is running firmware G.2.2. This should work on G.1.1 but I haven't tried it (though I'm pretty sure that's the version of firmware within which these directions were composed).

**Reply by eecue** [reply]
See my note below, I’m wondering if it’s related to my hardware differences. Also I have the latest Create firmware, I updated it as part of the initial setup.

**Reply by shamlian** [reply]
Also -- I am using a 2GB Nano image with version number jp461; I assume jp461 of the Nano (no suffix) image is the same version of Linux, but just to be sure:
```
jetson@nano:~$ uname -a
Linux nano 4.9.253-tegra #1 SMP PREEMPT Sat Feb 19 08:59:22 PST 2022 aarch64 aarch64 aarch64 GNU/Linux
```

**Comment by eecue** [comment]
One thing I noticed, in the hookup pictures it shows the Jetson Nano hooked up with two usb cables, however mine has a barrel jack where the second one is. I’m going to try attaching that directly to the 12V from the Create. Also going to try a few different usb cables.

**Reply by shamlian** [reply]
You have an original Jetson Nano; unfortunately, I don't have one of those to test. But looking at the Nano [docs](https://forums.developer.nvidia.com/t/power-supply-considerations-for-jetson-nano-developer-kit/71637), it seems like that is a 5 V input; I wouldn't attach that directly to the Create robot's VBATT output without a buck converter!! ☠️

**Reply by shamlian** [reply]
The hookup directions show a Nano 2GB, which is a confusingly-named variant with a different supply consideration.

**Reply by eecue** [reply]
Yeah was just about to look that up! I won't connect that to 12V!

**Reply by shamlian** [reply]
Reading the (original) Jetson Nano docs more closely, it seems like you should be able to get away with powering the original Nano with just a USB C (from Create 3) to USB Micro-B (to Nano) cable, as long as the Nano doesn't have any peripherals attached.

**Reply by wpd** [reply]
I doubt that changing the power supply wiring will make that much of a difference in the network connectivity issues you are having.  I wonder the Linux installed on your older Nano might inhibit ping responses... does it have any firewall rules created?

**Reply by eecue** [reply]
So just to double check, I built a USB to barrel jack cable and tried powering it via the barrel jack, same problem.   There are no firewall rules turned on.

**Comment by eecue** [comment, marked as answer]
Figured it out!!!! This is so weird, I was using a USB-C hub between the create and the Nano (per the instructions page) removing that and connecting the Nano directly to the Create fixed it!

**Reply by shamlian** [reply]
Awesome! Could you take a photo in the style of the other ones on the [hookup page](https://iroboteducation.github.io/create3_docs/hw/hookup/#nvidia-jetson-nanotm-2gb) (at your leisure) so we can add it to the docs? Alternatively, we'd be happy to accept a PR. That will help us to make it more explicit that there is a difference between the Nano and the Nano 2 GB. If not, that's OK; I should probably get ahold of an original Nano for testing.

**Reply by eecue** [reply]
I can yes. I don’t think that was explicitly my problem though. Pretty sure it was the hub I was using.

**Reply by shamlian** [reply]
Curious to know what the hub was doing to cause a problem. Would you be willing to share what make and model of hub you were using that seemed to interfere with the network connection?

**Comment by eecue** [comment]
Here’s the hub in question:

![094A9105-29C3-46E1-BFA6-F30F4F26AE13](https://user-images.githubusercontent.com/532536/173670164-6a3da783-300d-4957-835f-f4d1848db6f3.jpeg)

![AF0984C5-C47A-4A38-A319-4EC6F01E1A47](https://user-images.githubusercontent.com/532536/173670187-7fef41d3-a117-467a-bc6f-1a68cd35f6ae.jpeg)

It showed up normally but I couldn’t ping it. I could see traffic over tcpdump though.

**Reply by shamlian** [reply]
That's super weird. Thanks for the reporting and the help!

**Comment by ParadoxRobotics** [comment]
I think I have a similar issue...

I'm currently trying to connect my IRobot Create3 to a Jetson Nano 4Gb (Dev kit Rev A) : 
The jetson is powered using an external LiPo battery and connected to the robot using a USB C/USB A converter linked to a micro USB cable.

I followed the instruction in the documentation to set up my jetson but unfortunately I don't see the full topic list, only : 
```
/parameter_events
/rosout
```

Is the USB C/USB A converter the problem ?
What kind of USB C / micro USB cable do you use in you setup ?

**Reply by ParadoxRobotics** [reply]
The USB C/USB A converter was the problem...
